R/wrappers.R

```diff
@@ -0,0 +1,268 @@
+#'
+#' @importFrom httr GET
+#' @importFrom wrapr stop_if_dot_args
+#'
+#' @param sn A device serial number
+#' @param limit (optional) The number of devices to return
+#' @param sort (optional) A parameter upon which to sort, and the sort method (ascending or descending), formatted as "parameter,order", e.g. "id,asc"
+#'
+#' @returns The user's device information
+#' @export
+get_devices <- function(sn = NULL, limit = NULL, sort = NULL){
+  structure(
+    requests(paste("devices", sn, sep = "/"), verb = httr::GET, limit = limit, sort = sort),
+    class = "devices"
+  )
+}
+
+#' Coerce the devices class to a data.frame
+#'
+#' Converts the class returned by [get_devices()] to a data.frame.
+#'
+#' @importFrom magrittr `%>%`
+#' @importFrom tidyr unnest unnest_wider
+#' @importFrom dplyr rename_with
+#' @param x The devices class returned by [get_devices()]
+#' @param ... Placeholder for passing through to as.data.frame.default
+#' @returns A data.frame of device information
+#' @export
+as.data.frame.devices <- function(x, ...){
+  if(!is.null(names(x))){ # if x has names, then there's only one device returned
+    x <- list(x)
+  }
+
+  out_df <- do.call(rbind, lapply(x, rbind)) %>%
+    as.data.frame %>%
+    unnest_all() %>%
+    rename_with(~gsub("_1$", "", .x))
+
+  return(out_df)
+}
+
+#' Get device metadata
+#'
+#' @param sn Device serial number
+#' @returns The device metadata
+get_device_metadata <- function(sn){
+  structure(
+    requests(paste("meta-data", sn, sep="/")),
+    class = "metadata"
+  )
+}
+
+#' Get device data
+#'
+#' Get data according to provided serial number and other parameters.
+#'
+#' @importFrom httr GET
+#'
+#' @param sn A device serial number
+#' @param limit (optional) The number of data points to return
+#' @param start (optional) The earliest date to retrieve data from. Should be a timestamp string of the form "YYYY-MM-DD HH:MM:SS"
+#' @param stop (optional) The latest date to retrieve data from. Should be a timestamp string of the form "YYYY-MM-DD HH:MM:SS"
+#' @param filter (optional) A string providing filter parameters, see below examples and \href{https://docs.quant-aq.com/api#8e14edbf9dee4162a04f729ce022cb4b}{API documentation} for more information.
+#' @param sort (optional) A data variable upon which to sort, and the sort method (ascending or descending), formatted as "parameter,order", e.g. "timestamp,asc"
+#' @param raw (optional) Default FALSE. Returns the raw data. Currently only available to developers and admins.
+#'
+#' @examples
+#' \dontrun{
+#'   get_data(sn = "MOD-PM-00808", limit = 100)
+#'   get_data(sn = "MOD-PM-00808", start = "2020-01-01 21:54", stop = "2023-03-03 05:47")
+#'   get_data(sn = "MOD-PM-00808", filter = "device_state,eq,ACTIVE")
+#'   get_data(sn = "MOD-PM-00808", sort = "timestamp,asc")
+#' }
+#'
+#' @returns The specified device data
+#' @export
+get_data <- function(sn, limit = NULL, start = NULL, stop = NULL, filter = NULL, sort = NULL, raw = FALSE){
```

Comment:
I think we may want to set a non-NULL limit as right now the default behavior is to grab all data for a sensor (I think?) which can easily be millions of records which isn't something we likely want to encourage.